### ISSUE TITLE
Unit exporter `heading` is not actually units heading, it's the units 2d velocity vector

### ISSUE THREAD

**Issue by bitbubbler** [DCS-gRPC/rust-server#159]
https://github.com/DCS-gRPC/rust-server/blob/main/lua/DCS-gRPC/exporters/object.lua#L28-L30

The `heading` returned for a unit is only correct if the unit is moving exactly forward. This is because `getVelocity()` is being used to calculate the "heading". The result is actually a 2d "velocity vector", which is the correct velocity vector. If the unit is moving forward it should match or be very close to the units heading.

However, it is not the units heading. A unit moving backwards such as a helicopter will result in "heading" from the unit exporter equal to the direction the unit is moving.

The correct way to get the units heading appears to be with the `unit:getPosition()` function [(hoggit wiki)](https://wiki.hoggitworld.com/view/DCS_func_getPosition) using the same algo in use.

This is supported by this forum thread: https://forum.dcs.world/topic/148896-converting-a-units-orientation-to-a-quotheadingquot-script-help/

I'm able to confirm this with `CustomService.Eval` 
(I'm working in typescript, you can do the same in any language. Below code is a snippet)
```typescript
const lua = `return Unit.getByName("${name}"):getPosition()`

custom.eval({ lua }, (error, result) => {
  if (error) {
    console.log('failed to get unit position', error)
    return
  }
  const pos = JSON.parse(result!.json!)

  let heading = deg(Math.atan2(pos.x.z, pos.x.x))
  if (heading < 0) {
    heading = heading + 360
  }

  // here, heading value is always correct, even when moving backwards
})

function deg(radians: number): number {
  return radians * (180 / Math.PI)
}
```

warning: opinions below:

As a consumer of dcs-grpc, I want the units actual heading, In addition to the units velocity. (speed, velocity vector, etc)

I propose that instead of (or in addition to) extracting the heading and speed, expose the units raw velocity vec3. In addition to the units position vector 

There seems to be some attempt to expose the data available in lua outside of lua, but for units it doesn't align with the lua stuff very well. I had to figure out what `heading` was here. And currently have no access to raw vecs to do my own math

**Comment by rkusa** [comment]
Thanks for the report! I haven't really thought about it before, but now that I do, I'd personally expect the heading to be the direction a unit is moving to regardless of where the nose is pointing to.

> As a consumer of dcs-grpc, I want the units actual heading, In addition to the units velocity. (speed, velocity vector, etc)
> 
> I propose that instead of (or in addition to) extracting the heading and speed, expose the units raw velocity vec3. In addition to the units position vector
> 
> There seems to be some attempt to expose the data available in lua outside of lua, but for units it doesn't align with the lua stuff very well. I had to figure out what `heading` was here. And currently have no access to raw vecs to do my own math

We have an API to get information for units in 3D space: https://github.com/DCS-gRPC/rust-server/blob/650baa68ae39f77494f1f446fa72cb0d3675ae16/protos/dcs/unit/v0/unit.proto#L26-L30

It also includes, `orientation.yaw` which might be what you'd expect the heading to be?

**Comment by rurounijones** [comment]
I agree here that `heading` should be the direction the nose is pointing and `velocity` should be the direction the actual object is moving.

I think we didn't add velocity and other information in the `GetTransform` API because it would be extra work server-side and there was no use-case for it so it was work done for nothing. I have no objections to folding it into the stream.

@bitbubbler can you share the use-case?

**Comment by justin-lovell** [comment]
> it would be extra work server-side and there was no use-case for it so it was work done for nothing.

If there are performance bottlenecks being reported by server administrators, I do think there is merit in clients or server configurations to explicitly requesting additional data being attached to the stream decoration (including the `GetVelocity` call). 

That said, I don't think `GetTransform` would generate significant overhead to justify the above (until proven :) )

**Comment by bitbubbler** [comment]
Currently my case is for the units real nose position. I'm using the stream of unit positions to keep a database of "loosely accurate" unit positions (heading and velocity would be nice to have in this db)

I'm using the unit position and velocity in the database to do things like ctld and csar logic over grpc, so I Can't have to constantly poll the mission state (ctld and csar callouts don't require that much accuracy, a little delay is okay)

example: 
- grpc stream of units with custom poll intervals for unit types updates the db as soon as data is recieved (async)
- timer runs on the server for ever 1 second (javascript setInterval) (async)

timer runs the following:
- does a db read, and looks for all units near known csar locations (heavy on the db, but not heavy on the dcs server)
- do the math for clock callout (heading)
- do the math for eta (velocity and velocity vector)
- print `outTextForUnit` using the clock heading, distance, ETA.

I'm even able to do really cool things like determine that two units are converging the same csar, and ask one to wave off (multiplayer) once you have eta and velocity to a relative destination

In order to do all of this for everyone on a multiplayer server at once would require a ton of grpc calls. The stream solves this beautifully because in inverts the problem. The server already knows when units change, but it doesn't need to do any math.

I can also have my own trigger point outside of the game this way, I've got a few plans for the database of unit state. When you can put a few other threads to work you can even get pretty lazy about the math.

**Comment by bitbubbler** [comment]
Additional thoughts:

There is already options on the stream to control which types of events you get and how often you get them. I wonder if it would make sense to optionally ask for what type of unit data you get back? so one could "opt out" or "opt in" to the more expensive "Work" as part of the stream. Possibly even for a subset of units (say, just helicopters) as airplanes tend to travel forward

**Comment by rkusa** [comment]
> I agree here that `heading` should be the compass heading the nose is pointing and `velocity` should be the direction the actual object is moving.

I am not sure about that. Take the Apache as an example. When traveling around in the Apache, your nose is usually slightly of to one side (due to aerodynamics I think). So even while going full steam from A to B, heading and nose position aren't the same. But when I'd get asked by a flight member what my heading is, I wouldn't give them my nose position, I'd most definitly give them where I am heading/traveling to.

> I think we didn't add velocity and other information in the `GetTransform` API because it would be extra work server-side and there was no use-case for it so it was work done for nothing. I have no objections to folding that info into the stream.

`GetTransform` actually includes a velocity vector.

---

I am working on a proposal of integrating the information from `GetTransform` into each Unit export. While this work is motivated by my own use-cases, it would also fix some of the issues you encounter, e.g. the unit stream would include more information (like the velocity vector). I don't know if this proposal will land, but I'll reference it once the PR is up.

**Comment by rurounijones** [comment]
I think this is a case of "Technically correct" vs "general understanding." the helos certainly make this tricky but a plane will generally be pointing the direction it is heading +- a number of degrees that don't make a huge difference.

Similarly if you are in an Apache backpeddling most people would not, I think, say "Heading 180" when they are pointing north but would explicitly point out that they are "heading backwards 180" because it is such an unusual situation.

If we are going to get proper velocity vectors anyway then I think having the "Which way is the nose pointing" as a simple compass value integer would satisfy a lot of use-cases without having to get into the vector math.

**Comment by bitbubbler** [comment]
I feel like my example was partially missed here.

The only way to give a proper clock direction (3 o'clock, 6 o'clock) is to use the nose position. 
If you give a call of 9 o'clock you don't expect to look at 7/8 o'clock for a 9 o'clock  call out just because you're flying in an apache in aerodynamic flight.

same with cpg/pilot talk. you give clock heading based on nose position, always. there are no exceptions.

**Comment by rkusa** [comment]
I checked what TacView and DCS (on the F10 map) report as heading. Both report it as the direction the nose is pointing to (relative to the true north).

**Comment by rkusa** [comment]
Here is another fun one. I was the one defending heading as the direction of travel, though I was also the one implementing the current `GetTransform`, which I just noticed, [already returns the nose position](https://github.com/DCS-gRPC/rust-server/blob/3470157c2756b6daa5d997fd5d09492fbbbf1ffa/src/rpc/unit.rs#L88) rather than [the travel direction](https://github.com/DCS-gRPC/rust-server/blob/511af4bde32ac3559db5835dc1600698434e8208/lua/DCS-gRPC/exporters/object.lua#L35) 🤦‍♂️ - probably because I've noticed previously that this was the kind of heading that TacView expects ...

**Comment by rkusa** [comment]
Should be solved via #162. Units now have both `speed.heading` and `orientation.heading`.
